feat: Implement cache management for Monkeytype data

theg1239 opened the pull request:
- Added a centralized cache manager to handle fetching and caching of Monkeytype user data.
- Introduced background updates for cache every 5 minutes to prevent excessive API requests.
- Created API routes for cache monitoring and forced updates, accessible only to admin users.
- Updated admin panel to include a cache monitor link and refresh functionality.
- Enhanced error handling and logging for better debugging and user feedback.
- Refactored personal score update logic to utilize cached data and reduce unnecessary API calls.
- Improved leaderboard page to display cache statistics and automatic updates.

pk-vishnu commented:
Thanks for the PR ,really solid caching implementation! Just a few thoughts:

1. forceUpdate() behavior: Right now, it refreshes the cache for all users, even when just one user hits the update-personal-score endpoint. This could lead to unnecessary compute usage. Would it make sense for forceUpdate() to accept a userId and refresh only that specific user's cached data when triggered from that endpoint?

2. Balancing freshness and DB compute: To keep the experience responsive but avoid overloading the DB, maybe we can keep cache updates frequent (e.g., every 5 minutes), but limit actual DB writes to something like twice a day per user. That throttling could be handled inside the write function, by checking timestamps or a similar approach.

3. Cache statistics on the leaderboard: Could you clarify the point about "Improved leaderboard page to display cache statistics"? I didn’t see cacheStats being used in the current leaderboard view

theg1239 commented:
1. Implemented this change
2. Implemented it to make it twice per day for a user
3. That was something I removed but forgot that I did, ignore, was just a little footer with updates on when the cache was refreshed

pk-vishnu commented:
Awesome work! This is super close. Really appreciate the effort you’ve put in
Just a few minor TypeScript/linting issues that are breaking the build due to the strict config:

1. Admin cache view: The `error` variable on line 80 is defined but not used - either log/remove it
2. API route (update-leaderboard): NextRequest is imported but unused - `req `can be removed
3. To prevent cacheStats from being marked as unused, you might want to bring back the updates indicator in the footer
`{cacheStats && cacheStats.isUpdating ? " (Updating...)" : ""}` should do

Once those are sorted, should be good to merge!
